packages/flutter_tools/lib/src/ios/code_signing.dart

```diff
@@ -116,229 +159,787 @@ Future<Map<String, String>?> getCodeSigningIdentityDevelopmentTeamBuildSetting({
+    );
+    _logger.printStatus('[1]: ${_CodeSigningStyle.automatic.label} (recommended)');
+    _logger.printStatus('[2]: ${_CodeSigningStyle.manual.label}');
+    final String choice = await _terminal.promptForCharInput(
+      List<String>.generate(2, (int number) => '${number + 1}')..add('q'),
+      prompt: 'Select a signing style (or "q" to quit)',
+      defaultChoiceIndex: 0, // Just pressing enter chooses the first one.
+      logger: _logger,
+      displayAcceptedCharacters: false,
+    );
+    if (choice == '1') {
+      return _CodeSigningStyle.automatic;
+    } else if (choice == '2') {
+      return _CodeSigningStyle.manual;
+    }
+    return null;
+  }
+
+  /// Prompt user to select a code-signing identity from a list of [validCodeSigningIdentities].
+  /// Selects the first one found without prompting if there is no stdin or if
+  /// [autoSelectSingle] is true and only one identity was found.
+  ///
+  /// Throw an error if [throwOnCancel] is true and the user quits while
+  /// selecting an identity.
+  Future<String?> _selectSigningIdentity(
+    List<String> validCodeSigningIdentities, {
+    bool autoSelectSingle = false,
+    bool throwOnCancel = false,
+  }) async {
+    if (validCodeSigningIdentities.isEmpty) {
+      return null;
+    }
+
+    if (autoSelectSingle && validCodeSigningIdentities.length == 1) {
+      return validCodeSigningIdentities.first;
     }
 
     // If terminal UI can't be used, just attempt with the first valid certificate
     // since we can't ask the user.
-    if (!terminal.usesTerminalUi) {
+    if (!_terminal.stdinHasTerminal) {
       return validCodeSigningIdentities.first;
     }
+    _terminal.usesTerminalUi = true;
 
-    final int count = validCodeSigningIdentities.length;
-    logger.printStatus(
-      'Multiple valid development certificates available (your choice will be saved):',
+    _logger.printStatus(
+      '\nValid development certificates available (your choice will be saved):',
       emphasis: true,
     );
+    final int count = validCodeSigningIdentities.length;
     for (int i = 0; i < count; i++) {
-      logger.printStatus('  ${i + 1}) ${validCodeSigningIdentities[i]}', emphasis: true);
+      _logger.printStatus('[${i + 1}] ${validCodeSigningIdentities[i]}');
+    }
+    final String choice = await _terminal.promptForCharInput(
+      List<String>.generate(count, (int number) => '${number + 1}')..add('q'),
+      prompt: 'Please select a certificate for code signing (or "q" to quit)',
+      defaultChoiceIndex: 0, // Just pressing enter chooses the first one.
+      logger: _logger,
+      displayAcceptedCharacters: false,
+    );
+    if (choice == 'q') {
+      if (throwOnCancel) {
+        throwToolExit(
+          'No certificate was selected. Code signing is required to build a deployable iOS app.',
+        );
+      } else {
+        return null;
+      }
+    }
+    final String selectedCert = validCodeSigningIdentities[int.parse(choice) - 1];
```

Comment:
Should we check the choice was a valid number? Perhaps we should prompt the user again if the choice was invalid?

Reply:
Same comment for `_selectProvisioningProfile`

Reply:
Ahh snazzy, thanks for the link!